test/mpi/rma/putfence1.c

```diff
@@ -7,43 +7,54 @@
 #include "mpi.h"
 #include <stdio.h>
 #include <string.h>
+#include <limits.h>
 #include "mpitest.h"
-
-/* These counts allow reasonable sizes for the large tests */
-#define LARGE_CNT_CONTIG    550000000
-#define LARGE_CNT_NONCONTIG 150000000
+#include "dtpools.h"
 
 /*
 static char MTEST_Descrip[] = "Put with Fence";
 */
 
 static inline int test(MPI_Comm comm, int rank, int source, int dest,
-                       MTestDatatype * sendtype, MTestDatatype * recvtype)
+                       int i, int j, DTP_t send_dtp, DTP_t dest_dtp)
 {
     int errs = 0, err;
+    int len;
+    int disp_unit;
+    MPI_Aint sourcecount, destcount, count;
     MPI_Aint extent, lb;
     MPI_Win win;
+    MPI_Datatype sourcetype, desttype;
+    char send_name[MPI_MAX_OBJECT_NAME] = { 0 };
+    char recv_name[MPI_MAX_OBJECT_NAME] = { 0 };
+    void *sourcebuf, *destbuf;
+
+    count = send_dtp->DTP_type_signature.DTP_pool_basic.DTP_basic_type_count;
+    sourcetype = send_dtp->DTP_obj_array[i].DTP_obj_type;
+    desttype = dest_dtp->DTP_obj_array[j].DTP_obj_type;
+    sourcecount = send_dtp->DTP_obj_array[i].DTP_obj_count;
+    destcount = dest_dtp->DTP_obj_array[j].DTP_obj_count;
+    sourcebuf = send_dtp->DTP_obj_array[i].DTP_obj_buf;
+    destbuf = dest_dtp->DTP_obj_array[j].DTP_obj_buf;
+
+    MPI_Type_get_name(sourcetype, send_name, &len);
+    MPI_Type_get_name(desttype, recv_name, &len);
 
     MTestPrintfMsg(1,
-                   "Putting count = %ld of sendtype %s - count = %ld receive type %s\n",
-                   sendtype->count, MTestGetDatatypeName(sendtype), recvtype->count,
-                   MTestGetDatatypeName(recvtype));
-
-    /* Make sure that everyone has a recv buffer */
-    recvtype->InitBuf(recvtype);
-    MPI_Type_extent(recvtype->datatype, &extent);
-    MPI_Type_lb(recvtype->datatype, &lb);
-    MPI_Win_create(recvtype->buf, recvtype->count * extent + lb, extent, MPI_INFO_NULL, comm, &win);
+                   "Putting count = %ld of sourcetype %s - count = %ld receive type %s\n",
+                   sourcecount, send_name, destcount, recv_name);
+
+    MPI_Type_get_extent(desttype, &lb, &extent);
+    disp_unit = extent < INT_MAX ? extent : 1;
+    MPI_Win_create(destbuf, destcount * extent + lb, disp_unit, MPI_INFO_NULL, comm, &win);
```

Comment:
Why `disp_unit = extent < INT_MAX ? extent : 1;` is a solution for `large counts failure` ?